### PR TITLE
docs(String) add some documentation for String

### PR DESCRIPTION
**Pull request by rmosolgo**
I saw "Contributing.md" said you were interested in more docs for the standard library. 

I want to learn more about Crystal and I like writing docs so I thought I'd try. I've tried to take some clues from the surrounding documentation. What do you think of these?

Also, if you have the time, I have some questions:

- How do you like these docs? Is there anything I can do to make them better?
- How do you want to handle cases for methods with the same behavior but different signatures (eg, one accepts a `Char` and the other accepts a `String`, or one yields but the other returns an Array)? I see that the Ruby API docs group these methods, which seems good to me. I've often documented just one of the methods, but included the different signatures in the example.
- For future documentation, is there anything that you'd like documented more quickly? Or anything that should be left undocumented for now?

### CODE REVIEW

**Comment by asterite** [comment]
Thanks!

I think the time has come to decide how to document things. I usually prefer to use the 3rd person, as in "**Tests** whether `str` matches..." instead of "**Test** whether". I always imagine myself asking "What does this method do?" and the answer being "(it) **tests**...".

Opinions? We can include this decision in the `Contributing.md` so everyone follows this rule.

**Comment by rmosolgo** [comment]
:+1: for third person, it also matches `is_a?`, `responds_to?`, `includes?`, `starts_with?`

I'm happy to go back and update my docs & Contributing.md.

(By the way, THANK YOU for using third-person method names in the standard library, I think it's much better!)

File: src/string.cr [hunk 1]
```diff
@@ -1415,6 +1415,18 @@ class String
     cmp == 0 ? (bytesize <=> other.bytesize) : cmp
   end
 
+  # Test whether `str` matches `regex`.
```
Comment (jhass) [inline src/string.cr]: I think so far we use "\*str\* matches \*regex\*" to refer to argument names. In more places than not anyway.

File: src/string.cr [hunk 2]
```diff
@@ -1488,6 +1512,15 @@ class String
     end
   end
 
+  # Returns the index of `c` in the string, or `nil` if the string is not present.
```
Comment (jhass) [inline src/string.cr]: Maybe we should rename "c" to "char" here.

File: src/string.cr [hunk 3]
```diff
@@ -1694,6 +1750,20 @@ class String
     ary
   end
 
+  # Make an array by splitting the string on `separator` (and removing instances of `separator`).
+  # If `limit` is present, the array will be limited to `limit` items.
+  # The final item will contain the remainder of the string.
+  #
+  # If separator is a single space (`' '`, `" "`), it splits on any whitespace (see `#split()`).
+  #
+  # If the separator is `nil`, the string will be separated into one-character strings.
```
Comment (jhass) [inline src/string.cr]: Maybe mention to use `chars` instead though.

Comment (rmosolgo) [inline src/string.cr]: What do you mean by that?

Oh, I see an error here. To get single-character strings, you should pass an empty string:

```
~/projects/crystal $ crystal eval
puts
puts "Mississippi".split("")
^D
["M", "i", "s", "s", "i", "s", "s", "i", "p", "p", "i"]
```

**Comment by asterite** [comment]
I just added [this section](http://crystal-lang.org/docs/documenting_code/index.html) to the docs so we are all one the same page :-)

@rmosolgo Thanks for documenting this and wanting to document even more! Documentation is so, so important. It reduces the amount of questions made and answers needed, and makes everything look much more professional. So your help is really needed and appreciated.

File: src/string.cr (hunk 2 above)
Comment (rmosolgo) [inline src/string.cr]: I agree, I wasn't sure how much I should mess with existing code! Here's the refactor:

https://github.com/manastech/crystal/commit/29cab3169ad613baba601d5276e910dac69bb62f

**Comment by rmosolgo** [comment]
Thanks for your feedback, I've updated to follow that styleguide (very helpful, good idea!): 

https://github.com/rmosolgo/crystal/commit/321d5700a5592c73dd1363ce59f24fb2afe15337

Let me know what you think, or if it looks good, I'll squash them!

**Comment by jhass** [comment]
LGTM :+1:, squashing would be much appreciated ;)

**Comment by rmosolgo** [comment]
rebased!

**Comment by jhass** [comment]
Thanks!

**Pull request merged**